src/sender.ts

```diff
@@ -1,12 +1,109 @@
-import { Broadcast } from "./broadcast";
-import { Message } from "./message";
+import { logger } from "@blockr/blockr-logger";
+import { connect } from "socket.io-client";
+import { MessageType } from "./enums";
 
+import { Message } from "./models/message";
+
+/**
+ * Handles the sending of messages.
+ */
 export class Sender {
-    public sendMessage(message: Message) {
+    private readonly protocol = "http://";
+    private readonly port: string;
+    private readonly sentMessages: Map<string, Message>;
+    private readonly sentMessageSenders: Map<string, string>;
+
+    constructor(port: string) {
+        this.port = port;
+        this.sentMessages = new Map<string, Message>();
+        this.sentMessageSenders = new Map<string, string>();
+    }
 
+    /**
+     * Send the given message to the given destinationIp.
+     *
+     * @param message - The message
+     * @param destinationIp - The destinationIp ip
+     * @param destinationGuid
+     */
+    public sendMessage(message: Message, destinationIp: string, destinationGuid?: string): void {
+        this.emitMessage(message, destinationIp, destinationGuid);
     }
 
-    public sendBroadcast(broadcast: Broadcast) {
+    /**
+     * Send an acknowledge message to the given destinationIp, based on the given message.
+     *
+     * @param originalMessage - The message
+     * @param destinationIp - The destinationIp
+     */
+    public sendAcknowledgeMessage(originalMessage: Message, destinationIp: string): void {
+        const response = new Message(
+            MessageType.ACKNOWLEDGE,
+            originalMessage.originalSenderGuid,
+            originalMessage.guid,
+        );
+        response.createGuid();
+
+        // Send the response
+        const socket = connect(`${this.protocol}${destinationIp}:${this.port}`);
+        socket.emit("message", JSON.stringify(response));
+
+        logger.info(`Message sent to: ${destinationIp}: ${response.type}`);
+    }
+
+    /**
+     * Get guids from the senders of the sent messages after the given date.
+     * Deletes messages that are sent before the given date from the history.
+     *
+     * @param date - The date
+     *
+     * @return An array of the GUIDs of the senders
+     */
+    public getSentMessagesSendersSince(date: Date): string[] {
+        const guids: string[] = [];
+
+        this.sentMessages.forEach((value: Message, key: string) => {
```

Comment:
Use the `for of` loop instead. As the `foreach` creates a new closure that might be unwanted.
The `for of` loop is also more efficient for the typescript compiler.